File: go/internal/tools/clock.go
package tools

import (
	"context"
	"time"
)

// clockCtxKey namespaces the reference clock on a context. Its own unexported
// type is what keeps the value from colliding with another package's.
type clockCtxKey struct{}

// WithClock attaches the reference clock the audit report resolves a relative
// since_offset against. The cross-language behavior fixtures pin a report whose
// window is relative, and a wall-clock window has no reproducible answer.
func WithClock(ctx context.Context, clock func() time.Time) context.Context {
	if clock == nil {
		return ctx
	}

	return context.WithValue(ctx, clockCtxKey{}, clock)
}

// ClockFromContext answers the attached reference time, or the wall clock when
// the caller attached none.
//
// Exported because the generated handler reads it: LOCAL_AMBIENT_CLOCK renders
// in this language as the expression the call site hands its operation.
func ClockFromContext(ctx context.Context) time.Time {
	clock, ok := ctx.Value(clockCtxKey{}).(func() time.Time)
	if !ok {
		return time.Now()
	}

	return clock()
}

File: python/src/linodemcp/tools/clock.py
"""Ambient reference clock for the audit report tool.

Python tool handlers take ``(arguments, config)`` with no context parameter, so
the reference clock travels through a ``ContextVar`` the way the builder state
does (``linodemcp.tools.builderstate``). Go carries the same seam on the call
context (``WithClock`` in ``go/internal/tools/clock.go``).
"""

from __future__ import annotations

from contextvars import ContextVar
from datetime import UTC, datetime
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from collections.abc import Callable
    from contextvars import Token

_CLOCK: ContextVar[Callable[[], datetime] | None] = ContextVar(
    "reference_clock", default=None
)


def set_clock(
    clock: Callable[[], datetime] | None,
) -> Token[Callable[[], datetime] | None]:
    """Publish the reference clock, returning a resetting token.

    The cross-language behavior fixtures pin a report whose window is relative,
    and a wall-clock window has no reproducible answer.
    """
    return _CLOCK.set(clock)


def reset_clock(token: Token[Callable[[], datetime] | None]) -> None:
    """Restore the clock the matching set_clock call replaced."""
    _CLOCK.reset(token)


def now() -> datetime:
    """Return the published reference time, or the wall clock when unset."""
    clock = _CLOCK.get()
    if clock is None:
        return datetime.now(UTC)

    return clock()
